pkg/controller/deployment/recreate.go

```diff
@@ -131,6 +139,38 @@ func (dc *DeploymentController) waitForInactiveReplicaSets(oldRSs []*extensions.
 	return nil
 }
 
+// waitForNoPods will wait until all pods for the provided deployment are deleted.
+func (dc *DeploymentController) waitForNoPods(deployment *extensions.Deployment) error {
+	selector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return err
+	}
+	pods, err := dc.podLister.Pods(deployment.Namespace).List(selector)
+	if err != nil {
+		return err
+	}
+	if len(pods) == 0 {
+		return nil
+	}
+	options := api.ListOptions{LabelSelector: selector, ResourceVersion: pods[0].ResourceVersion}
+	w, err := dc.client.Core().Pods(deployment.Namespace).Watch(options)
+	if err != nil {
+		return err
+	}
+	defer w.Stop()
+
+	deletionsNeeded := len(pods)
+	condition := func(event watch.Event) (bool, error) {
+		if event.Type == watch.Deleted {
+			deletionsNeeded--
+		}
+		return deletionsNeeded == 0, nil
+	}
+	// TODO: Wait some time proportionate to the size of deletionsNeeded.
+	_, err = watch.Until(2*time.Minute, w, condition)
```

Comment:
Nothing. The controller would start recreating the new pods as soon as all the old pods had been marked deleted. Normally, this timeout should be infinite, Deployments that use progressDeadlineSeconds should have pds-(now()-lastUpdateTime[1]).

Will update

[1] Time of the Progressing condition